### PR TITLE
[Form] added `CallbackChoiceLoader` and refactored ChoiceType's children

### PR DESCRIPTION
**Pull request by HeahDude**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | ~ |
| License | MIT |
| Doc PR | todo |
# Todo
- [ ] Address a doc PR
- [x] Update CHANGELOG
# Changes
- 39e937f added `CallbackChoiceLoader` to lazy load choices with a simple callable.
- 995dc56 refactored `CountryType`, `CurrencyType`, `LanguageType`, `LocaleType` and `TimezoneType` for better performance by implementing `ChoiceLoaderInterface` for lazy loading.
# Usage

``` php
use Symfony\Component\Form\ChoiceList\Loader\CallbackChoiceLoader;
use Symfony\Component\Form\Extension\Core\Type\ChoiceType;

$builder->add('constants', ChoiceType::class, array(
    'choice_loader' => new CallbackChoiceLoader(function() {
            return StaticClass::getConstants();
    },
));
```

### CODE REVIEW

**Comment by HeahDude** [comment]
Basic usage:
```php
$builder->add('locales', LocaleType::class, array(
    'choice_filter' => function ($choice) use ($userLocales) {
        return in_array(key($choice), $userLocales);
    },
));

$builder->add('entities', 'EntityType::class, array(
    'class' => \AppBundle\Entity\Issue::class,
    'choice_filter' => function ($choice) use ($userName) {
        return $userName === $choice->getAssignee();
    },
));
```

File: src/Symfony/Component/Form/Extension/Core/Type/CurrencyType.php [hunk 1]
```diff
@@ -11,36 +11,42 @@
...
+     * currency fields on the same page don't lead to
+     * unnecessary overhead.
+     *
+     * @return array The currency choices
      */
-    public function getBlockPrefix()
+    protected function loadChoices()
     {
-        return 'currency';
+        if (null === self::$currencies) {
```
Comment (stof) [inline src/Symfony/Component/Form/Extension/Core/Type/CurrencyType.php]: storing them in a static property will increase the memory usage by preventing garbage collection until the end of the process.
What kind of performance gain is there in term of time here, to counter-balance the loss regarding memory perf ?

Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/CurrencyType.php]: Those types may be used many times in a form or with many forms in one request.

This optimisation is already done in `TimeZoneType`.

I've have no hard feeling about this, just a PoC for now.

File: src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php [hunk 2]
```diff
@@ -150,6 +150,38 @@ public function createListFromLoader(ChoiceLoaderInterface $loader, $value = nul
...
+     *
+     * The callable should be compatible with array_filter and
+     * return true to keep a loaded choice or false to discard it.
+     *
+     * @param ChoiceListInterface $choiceList The loaded choice list to filter
+     * @param callable            $filter     The filter to apply to the choices
+     *
+     * @return ChoiceListInterface The filtered choice list
+     */
+    public function filterChoiceList(ChoiceListInterface $choiceList, callable $filter)
```
Comment (stof) [inline src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php]: Adding this method to this class is wrong, as this class is only a particular implementation of ChoiceListFactoryInterface and ChoiceType must accept any implementation

Comment (stof) [inline src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php]: btw, I think this logic should not be in the factory at all. It is a totally different business logic than creating choice lists, which becomes clear when removing duplication in your code:

```php
$choices = array_filter($choiceList->getChoices(), $filter);

return $this->createListFromChoices($choices);
```

This means that the filtering should be moved to a consumer of the factory, which would then call the public API of the factory (supporting any factory, not only the caching one)

Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php]: I've thought about adding a new Interface and this method in all the factories to not break BC, but use this as "short cut" (PoC) otherwise it may not be implementable at all...

File: src/Symfony/Component/Form/Extension/Core/Type/CurrencyType.php (hunk 1 above)
Comment (stof) [inline src/Symfony/Component/Form/Extension/Core/Type/CurrencyType.php]: Sure they can be used many times. But any such optimization like this must be backed by numbers (especially when it is expected to improve one side at the expense of a loss on the other side, to know whether the loss can be accepted)

Comment (stof) [inline src/Symfony/Component/Form/Extension/Core/Type/CurrencyType.php]: anyway, I would suggest splitting this caching outside this PR, to discuss it separately (this caching can even be applied without the refactoring done in this PR)

Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/CurrencyType.php]: Ok. Thanks @stof for the review!

Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/CurrencyType.php]: @stof this PR is now about the caching in a first commit and a refactoring in a second commit.

For `choice_filter` and a better approach in the implementation see #18334.

I would very much appreciate some feedback there before adding more tests. Thanks.

**Comment by HeahDude** [comment]
Ok so I dropped 8298b1d because the `CachingFactoryDecorator` (ref #16747) already do the job, then there is no gain in caching the `array_flip` even if it is justified for `TimezoneType`.

I still think this abstract class is good for DX and maintainability. For example introducing the `choice_translation_domain` option would have required one line to this abstract class instead of modifying 5 classes.

This may be trivial but there is no harm in performance, I've tried this code in SE:
```php
$builder = $this->createFormBuilder();

for ($i = 0; $i < 10; ++$i) {
    $builder->add('country'.$i, FormType\CountryType::class);
    $builder->add('currency'.$i, FormType\CurrencyType::class);
    $builder->add('language'.$i, FormType\LanguageType::class);
    $builder->add('locale'.$i, FormType\LocaleType::class);
    $builder->add('timezone'.$i, FormType\TimezoneType::class);
}
$form = $builder->getForm();
```
From base to 8298b1d (try to improve cache with static properties:
10_form_base_to_cache

And from base to refactoring (current PR):
10_from_base_to_refactor

**Comment by HeahDude** [comment]
ping @webmozart, does it fit with #18368 ?

**Comment by webmozart** [comment]
A simpler and more flexible solution would be to add a `CallableChoiceLoader`. To the constructor of that loader, we can pass a callable in each of these types that loads the choices, e.g.:

~~~php
class CountryType extends AbstractType
{
    private $choiceLoader;

    public function __construct()
    {
        $this->choiceLoader = new CallableChoiceLoader(function () {
            return array_flip(Intl::getRegionBundle()->getCountryNames());
        });
    }

    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults([
            'choice_loader' => $this->choiceLoader,
        ]);
    }
}
~~~

For each type, we make sure to reuse the same loader instance, then the caching mechanisms in `ChoiceType` will automatically take effect.

**Comment by HeahDude** [comment]
I thought about it too since it is very easy to do, I'm glad you brought that up, I might be done in time :)

Should I go on with it ?

**Comment by stof** [comment]
I prefer the solution suggested by @webmozart

**Comment by HeahDude** [comment]
Rebased.

I first thought of using `choice_loader` as any callable when dealing with #18334, but I was not sure that it would be good to add an interface for that, although that's looks very great for those sub types!

I think this callable choice loader will really improve performances, I will try to profile something today :)

Also, I still think of some things:
 - it would be good to decouple the need to inherit from choice type basis from the need to load choices in a permanent way to ease maintaining them.

 This was originally sent as PoC, but it would be good to find a better name for such an abstract class:
  `LoadedChoiceType` or `AbstractLazyChoiceType`, it would be a more generic core version of `DoctrineType`.

- I feel `TimezoneType` is looking really good now but the four others could be more optimised soon (in regard of #18368) to directly implement the `ChoiceLoaderInterface` as we can do for mappers or transformers (e.g #18357).


Concerning choice loaders in general, if this PR is merged, this would be the first one implemented in the form core and in regard of the work I'm trying to do in #18359, I think there is two much caching when using loaders here's a schema:
```
Needs choice list (configuration | loading form types)
|
=> factory
   |
   |__ from choices ?
   |   |
   |   |__ decorated with cache ?
   |   |   |      
   |   |   |__ => return ArrayChoiceList
   |   |
   |   |
   |   |__ decorates
   |   |
   |   |__ => creates
   |       |
   |       |__ => cache => return ArrayChoiceList 
   |
   |
   |__ from loader ?
	   |
       |__ decorated with cache ?
           |  
           |__ => return ArrayChoiceList
		   |
		   |
		   |__ decorates
		       |
			   |__ => cache (*1) and return **LazyChoiceList**


Then later or not uses choice list (actually build form or view)
|
=> not lazy ?
    |
	|__ => return classic (good)
	|
	|__ is loaded ?
	    |
		|__ => return classic
		|
		|__ call loader
		    |
			|__ is loaded ?
			|   |
			|   |__ => return classic
			|   |
			|   |__ call factory
			|   |   |
			|   |   |__ decorated with cache
			|   |   |    |
			|   |   |    |__ wtf ?
			|   |   |
			|   |   |__ decorated with property path ?
			|   |   |    |
			|   |   |    |__ wtf ?
			|   |   |
			|   |   |__ factory creates, cache (*2) and return ArrayChoiceList
			|   |
			|   |__ loader cache (*3)
			|
			|__ lazy choice list cache (*4)
```

The `(*n)` marker indicates when the caching implies the number of copy of the same choice list.

So I suggest to use the following code in `LazyChoiceList`:
```
    public function getChoices()
    {
        // BC
        if (!$this->isLoaded) {
            $this->loadedList = $this->loader->loadChoiceList($this->value);
            $this->isLoaded = true;
        } else {
            if ($this->loadedList !== $loadedList = $this->loader->loadChoiceList($this->value) {
                // Trigger deprecation notice for caching in lazy choice list
                // not caching in choice loader unsupported in 4.0
                return $this->loadedList->getChoices();
            }
        }
        // End bc
        return $this->loader->loadChoiceList($this->value)->getChoices();
    }
```

And together with the change of #18359 we can get in 4.0:
```
Uses choice list (actually build form or view)
|
=> not lazy ?
    |
	|__ => return classic (good)
	|
	|
	|__ call loader
		|
		|__ is loaded ?
	        |
		    |__ => return classic
		    |
		    |__ loader creates, cache (*2) and returns ChoiceListInterface
```

I think the loader should not know about factories. The `DoctrineChoiceLoader` is not registered as service, it is only instantiated by default in the `DoctrineType` which passes it the factory so it's not really an extension point.

What do you think ?

Thanks.

File: src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        return $this->choiceList = new ArrayChoiceList(call_user_func($this->callback), $value);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadChoicesForValues(array $values, $value = null)
+    {
+        // Optimize
+        if (empty($value)) {
```
Comment (webmozart) [inline src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php]: `$values`? Is this not tested?

File: src/Symfony/Component/Form/Extension/Core/Type/ImmutableChoiceType.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * Base class for child of {@link ChoiceType}.
+ *
+ * Every class extending it is responsible for loading
+ * the choices and eventually cache them.
+ *
+ * @author Jules Pietri <jules@heahprod.com>
+ */
+abstract class ImmutableChoiceType extends AbstractType
```
Comment (webmozart) [inline src/Symfony/Component/Form/Extension/Core/Type/ImmutableChoiceType.php]: I don't think this class is needed.

File: src/Symfony/Component/Form/Extension/Core/Type/CountryType.php [hunk 5]
```diff
@@ -11,36 +11,33 @@
...
-    }
-
-    /**
-     * {@inheritdoc}
-     */
-    public function getParent()
+    public function getBlockPrefix()
     {
-        return __NAMESPACE__.'\ChoiceType';
+        return 'country';
```
Comment (webmozart) [inline src/Symfony/Component/Form/Extension/Core/Type/CountryType.php]: Why did you add this? The block prefix is "country" by default.

File: src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php (hunk 3 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php]: Typo...

File: src/Symfony/Component/Form/Extension/Core/Type/CountryType.php (hunk 5 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/CountryType.php]: You should switch in split view :)

File: src/Symfony/Component/Form/Extension/Core/Type/ImmutableChoiceType.php (hunk 4 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/ImmutableChoiceType.php]: It avoids a lot of code duplication IMHO and could be good for DX.

If you consider `getBlockPrefix()`  is useless each type would be responsible only for implementing the abstract method only.

It feels like a good decoupling to me.

However I'm absolutely not against removing it :)

Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/ImmutableChoiceType.php]: Also, it allows to write tests for all sub types without being skipped by ICU condition.

File: src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php (hunk 3 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php]: The choice loader should not be responsible for `$value` check or test IMHO.

File: src/Symfony/Component/Form/Extension/Core/Type/ImmutableChoiceType.php [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+     * {@inheritdoc}
+     */
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        $resolver->setDefaults(array(
+            'choice_loader' => $this->getChoiceLoader(),
+            'choice_translation_domain' => false,
+        ));
+
+        $resolver->setRequired(array('choice_loader'));
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/ImmutableChoiceType.php]: I guess it's better to be strict in an abstract class here and `getChoiceLoader` as abstract looks safer to maintain.

**Comment by HeahDude** [comment]
When exactly will 3.1 be feature frozen ?

**Comment by fabpot** [comment]
End of the month :) Old PR can be merged next week at the latest, new ones, no.

**Comment by HeahDude** [comment]
@fabpot the underlying question: can the first commit (just adding the `CallbackChoiceLoader` feature) be merged now as its ready, and the internal "enhancement/refactoring" of the sub choice types be worked on during the beta ?

**Comment by HeahDude** [comment]
OK I've got your message after typing the second :) so no need to rush!

**Comment by HeahDude** [comment]
Thanks

File: src/Symfony/Component/Form/Extension/Core/Type/ImmutableChoiceType.php (hunk 4 above)
Comment (webmozart) [inline src/Symfony/Component/Form/Extension/Core/Type/ImmutableChoiceType.php]: I think the benefits of adding this abstract class don't outweigh the disadvantages of a more complex class inheritance tree. Usually composition in types should be preferred over inheritance, since inheritance is much less flexible.

For example, if you add a protected method and later find out you need another argument in that method, your limited by our BC policy. Creating a new object instance within a method body is not restricted by that policy and much more flexible.

File: src/Symfony/Component/Form/Extension/Core/Type/CountryType.php (hunk 5 above)
Comment (webmozart) [inline src/Symfony/Component/Form/Extension/Core/Type/CountryType.php]: I don't understand what you're trying to say. The default return value of that method is "country". This method seems superfluous to me.

File: src/Symfony/Component/Form/Extension/Core/Type/ImmutableChoiceType.php (hunk 4 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/ImmutableChoiceType.php]: OK, thanks for this overview :)

File: src/Symfony/Component/Form/Extension/Core/Type/CountryType.php (hunk 5 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/CountryType.php]: I mean you might be looking at the diff "unified", and should try "split", because this method is not changed at all :)

**Comment by HeahDude** [comment]
Comments addressed, rebased, updated title and description.

Looks really good now! I'll try to send blackfire profiles soon!

**Comment by HeahDude** [comment]
@webmozart I still think it can be improved for all classes excepted `TimezoneType` by letting them implementing `ChoiceLoaderInterface` directly once #18334 dealt with:
```php
<?php

/*
 * This file is part of the Symfony package.
 *
 * (c) Fabien Potencier <fabien@symfony.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Symfony\Component\Form\Extension\Core\Type;

use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\ChoiceList\Loader\ChoiceLoaderInterface;
use Symfony\Component\Intl\Intl;
use Symfony\Component\OptionsResolver\OptionsResolver;

class CountryType extends AbstractType implements ChoiceLoaderInterface
{
    /**
     * The loaded countries.
     *
     * @var ArrayChoiceList
     */
    private $countries;

    /**
     * {@inheritdoc}
     */
    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults(array(
            'choice_loader' => $this,
            'choice_translation_domain' => false,
        ));
    }

    /**
     * {@inheritdoc}
     */
    public function getParent()
    {
        return __NAMESPACE__.'\ChoiceType';
    }

    /**
     * {@inheritdoc}
     */
    public function getBlockPrefix()
    {
        return 'country';
    }

    public function loadChoiceList()
    {
        if ($this->countries) {
            return $this->countries;
        }

        $countries = array_flip(Intl::getRegionBundle()->getCountryNames());

        return $this->countries = new ArrayChoiceList($countries);
    }

    public function loadChoicesForValues($values, $value)
    {
        // Optimize, as already done in this PR
        if (empty($values) {
            return array();
        }

        if ($this->countries) {
            return $this->countries->getChoicesForValues($values);
        }

        // Optimisation not implement in this PR requiring new API in \Intl component
        return Intl::getRegionBundle()->getCountriesWithNames($values);
    }

    // ...
```

What do you think ?

**Comment by HeahDude** [comment]
I've made some profiling. Symfony standard, a simple form with a field for each optimised type involved in that PR:

- [on initial display](https://blackfire.io/profiles/compare/921fa862-70bd-42a3-bbf6-7e04ebf9cec5/graph)

init_immutable

- [on cached display](https://blackfire.io/profiles/compare/81181fc8-b11c-48a5-bf99-f7ec5d6c8b31/graph)

display_immutable

- [on submit](https://blackfire.io/profiles/compare/f8a2da88-542f-4919-a9a5-ab0511da0711/graph)

submit_immutbale

Check the real numbers, they are cool :)

Here's the test on symfony-standard:
```php
/**
 * @Route("/", name="homepage")
 */
public function indexAction(Request $request)
{
    // Load fixtures
    //$this->warmUp();

    $builder = $this->createFormBuilder()
        ->add('country', '\Symfony\Component\Form\Extension\Core\Type\CountryType')
        ->add('currency', '\Symfony\Component\Form\Extension\Core\Type\CurrencyType')
        ->add('language', '\Symfony\Component\Form\Extension\Core\Type\LanguageType')
        ->add('locale', '\Symfony\Component\Form\Extension\Core\Type\LocaleType')
        ->add('timezone', '\Symfony\Component\Form\Extension\Core\Type\TimezoneType')
        ->add('Save', '\Symfony\Component\Form\Extension\Core\Type\SubmitType')
    ;

    // Start profiling
    $probe = $this->createBlackFireProbe('PR'.($submit ? ' (submit)' : ''));

    // Loads the types
    $form = $builder->getForm();

    // On submit load the choice by callback
    $form->handleRequest($request);

    // Load the choices by callback if needed
    $view = $form->createView();

    // End profiling
    $this->sendProbe($probe);

    return $this->render('default/index.html.twig', [
        'base_dir' => realpath($this->getParameter('kernel.root_dir').'/..'),
        'form' => $view,
    ]);
}
```

**Comment by HeahDude** [comment]
That would be nice to get final reviews so it gets in 3.1. Thanks!

**Comment by HeahDude** [comment]
This PR enhancements are also impacted with the lazy choice list optimized from #18359 compared to current 3.1 we got:

 - [on display](https://blackfire.io/profiles/compare/cbf990be-622c-4e84-ad4c-25e1f94ce667/graph)

bonus_display

 - [on submit](https://blackfire.io/profiles/compare/547924a8-0d0c-48f4-a106-ad20a6e3952e/graph)

bonus_submit


And in 4.0 without BC layer compared to current 3.1:

 - [on display](https://blackfire.io/profiles/compare/114de490-ed93-43b2-82a4-cfc595881252/graph)

init_4

 - [on submit](https://blackfire.io/profiles/compare/7731289f-35a8-4c1b-b35f-99bb683cb495/graph)

submit_4

**Comment by HeahDude** [comment]
Maybe we should also define the default `choice_value` to:
```php
function ($choice) { return $choice;}
```
to prevent the `castableToString` being called, it's not needed here, what do you think @webmozart ?

File: src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+     */
+    public function loadChoicesForValues(array $values, $value = null)
+    {
+        // Optimize
+        if (empty($values)) {
+            return array();
+        }
+
+        if (null === $this->choiceList) {
+            $this->loadChoiceList($value);
```
Comment (stof) [inline src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php]: this could be written as ``$this->loadChoiceList($value)->getChoicesForValues($values);`` as ``loadChoiceList`` is responsible for returning the ChoiceList, instead of relying on the method side-effect of initializing its memoizing property

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php [hunk 8]
```diff
@@ -51,38 +60,33 @@ public function getBlockPrefix()
...
-     * so multiple timezone fields on the same page don't lead to unnecessary
-     * overhead.
+     * Returns a normalized array of timezone choices.
      *
      * @return array The timezone choices
+     *
+     * @internal This method is public to be used as callback and is not meant for user-land code.
      */
-    private static function getTimezones()
+    public static function getTimezones()
```
Comment (stof) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: couldn't we use a closure accessing to the private method to keep this private ?

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/LazyChoiceTypeTest.php [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        $this->assertContains(new ChoiceView('en', 'en', 'English'), $choices, '', false, false);
+        $this->assertContains(new ChoiceView('en_GB', 'en_GB', 'English (United Kingdom)'), $choices, '', false, false);
+        $this->assertContains(new ChoiceView('zh_Hant_MO', 'zh_Hant_MO', 'Chinese (Traditional, Macau SAR China)'), $choices, '', false, false);
+    }
+}
+
+//*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/*/
+//
+// Fixtures
+//
```
Comment (stof) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/LazyChoiceTypeTest.php]: This should be removed

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/LazyChoiceTypeTest.php [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+use Symfony\Component\Form\Test\TypeTestCase as TestCase;
+use Symfony\Component\Form\ChoiceList\View\ChoiceView;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * Tests children of choice type using a default choice loader.
+ *
+ * @author Jules Pietri <jules@heahprod.com>
+ */
+class LazyChoiceTypeTest extends TestCase
```
Comment (stof) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/LazyChoiceTypeTest.php]: What is this test ? there is no LazyChoiceType class

File: src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php (hunk 7 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php]: :+1: Nice catch, thanks.

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php (hunk 8 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: Good idea, work on that, thanks.

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/LazyChoiceTypeTest.php (hunk 10 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/LazyChoiceTypeTest.php]: I hope you read below :) It's about testing the refactored choice types, to test the loading and since their current tests are (often, almost always :) skipped because of ICU conflict.

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/LazyChoiceTypeTest.php (hunk 9 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/LazyChoiceTypeTest.php]: will do

**Comment by HeahDude** [comment]
Thanks @stof for that review :)

**Comment by HeahDude** [comment]
Comments addressed, rebased on master, tests green.

Final round of reviews and maybe a chance to get this in 3.1 ? Thanks.

File: src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php (hunk 7 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php]: In fact, I realised something we can optimize those two methods a lot. For all these types using ICU libraries, we only need to load them to get the proper localized labels, because unless `choice_value` is explicitly set, the choices are the same as the values.

Hence, when `$value` is `null` we just need to return the values!
And same goes for `loadValuesForChoices`, we just need to return the choices.

I will update one final time this PR, cool :)

File: src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+     */
+    private $callback;
+
+    /**
+     * The loaded choice list.
+     *
+     * @var ArrayChoiceList
+     */
+    private $choiceList;
+
```
Comment (stof) [inline src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php]: please add documentation about what the callable is expected to return (namely an array of choices)

File: src/Symfony/Component/Form/Tests/ChoiceList/Loader/CallbackChoiceLoaderTest.php [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+{
+    protected function setUp()
+    {
+        parent::setUp();
+    }
+
+    protected function tearDown()
+    {
+        parent::tearDown();
+    }
```
Comment (stof) [inline src/Symfony/Component/Form/Tests/ChoiceList/Loader/CallbackChoiceLoaderTest.php]: These are useless as they don't do anything except calling the parent one

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/LazyChoiceTypeTest.php (hunk 10 above)
Comment (stof) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/LazyChoiceTypeTest.php]: Well, there is strictly nothing testing the runtime code of the CountryType there, as the CountryType does not run.
Testing that fixtures classes not used anywhere else are working properly is useless, as nobody will ever use them anyway.

And if the goal is to test the usage of the CallbackChoiceLoader, this should be done in the tests of the choiceType or of the CallbackChoiceLoader (depending of what they test).

Comment (stof) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/LazyChoiceTypeTest.php]: and btw, AppVeyor runs the right version of ICU, so all PRs are tested properly on our CI.

**Comment by HeahDude** [comment]
@stof All comments addressed here, plus the optimization by implementing the `ChoiceLoaderInterface` directly!

This one still has a chance. I'm just waiting for the appveyor tests to be green :)

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php [hunk 13]
```diff
@@ -51,38 +55,79 @@ public function getBlockPrefix()
...
+
+        return $this->loadChoiceList($value)->getValuesForChoices($choices);
+    }
+
+    /**
+     * Returns a normalized array of timezone choices.
      *
      * @return array The timezone choices
      */
     private static function getTimezones()
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: this doesn't have to be static anymore, right ?

**Comment by HeahDude** [comment]
Green :)

**Comment by HeahDude** [comment]
I will try to profile again but there's no doubt that they'll be better than above :)

File: src/Symfony/Component/Form/Tests/ChoiceList/Loader/AbstractChoiceLoaderTest.php [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+ */
+
+namespace Symfony\Component\Form\Tests\ChoiceList\Loader;
+
+use Symfony\Component\Form\ChoiceList\LazyChoiceList;
+
+/**
+ * @author Jules Pietri <jules@heahprod.com>
+ */
+abstract class AbstractChoiceLoaderTest extends \PHPUnit_Framework_TestCase
```
Comment (stof) [inline src/Symfony/Component/Form/Tests/ChoiceList/Loader/AbstractChoiceLoaderTest.php]: IMO, there is no need to introduce this abstract class, as you use it only once

Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/ChoiceList/Loader/AbstractChoiceLoaderTest.php]: Don't you think it might help to introduce other loaders in the future and test them ?
I did it because it's the first implementation in the form component (the only one before was the `DoctrineChoiceLoader`) but this part should be re-usable IMHO.

Of courses I can easily integrate it in `CallbackChoiceLoaderTest` class.

Is this the final blocker ? Thank you again for your review.

Comment (xabbuh) [inline src/Symfony/Component/Form/Tests/ChoiceList/Loader/AbstractChoiceLoaderTest.php]: @HeahDude We can still extract common logic in an abstract base class then if needed. I would remove this class for now too.

Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/ChoiceList/Loader/AbstractChoiceLoaderTest.php]: Fair enough! I do it right away.

**Comment by HeahDude** [comment]
Alright, last comments addressed! Failures are unrelated, I think I'm done here :)

**Comment by fabpot** [comment]
👍

**Comment by HeahDude** [comment]
Rebased. Tests still green.

Last comment here is @fabpot approval. Is this still planned for 3.1? If so, it should be merged as soon as possible to be tested during beta.

ping __@symfony/deciders__

Thanks.

**Comment by HeahDude** [comment]
Rebased.

**Comment by HeahDude** [comment]
Re-re-rebased :)

File: src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php (hunk 11 above)
Comment (fabpot) [inline src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php]: @HeahDude You seemed to agree, but I don't see any documentation here. Can you add a phpdoc? I would then remove the doc on the property above.

Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php]: @fabpot Sorry, I don't understand where you want me to add and remove things.

Back then I had changed the first phpdoc from `The callable used to load the choices.` to `The callable used to load the array of choices.` thinking that both phpdocs would be less confusing.

Comment (fabpot) [inline src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php]: Sorry, I was not very clear. Let me try again. It would be better to add a phpdoc to the constructor:

```php
 /**
  * @param callable $callback The callable used to load the array of choices
  */
public function __construct(callable $callback)
 {
     $this->callback = $callback;
}
```

and remove the one on the private property:

```php
private $callback;
```

Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php]: Alright, makes sense. I will do it very soon.

**Comment by HeahDude** [comment]
Comments addressed and changelog updated to target 3.2.

**Comment by fabpot** [comment]
@HeahDude Can you submit a PR for the docs? Thanks.

**Comment by fabpot** [comment]
Thank you @HeahDude.

**Comment by HeahDude** [comment]
Will do, thank you @fabpot
